Fix invalid foreach in SharingBlacklist.php

phil-davis opened the pull request:
## Description
The "string" value that it is set to is supposed to be valid JSON representing an array of group ids.

`blacklisted_receiver_groups` can be set "manually" by the administrator using an `occ` command, or by some automated test or... So it can end up being an empty string, or some string that is not valid JSON, or some string that represents a multi-level nested array-of-arrays, or some other weird structure.

1) If the JSON decode does not produce an array, then return an empty list. (JSON decode might have failed, or returned a plain string/int etc)

2) Check the elements of the decoded array. Keep elements that are "string-like" (string, int and float) and cast them to string. Discard other elements (e.g. an array, or boolean). This allows for group names like "0", "1", "1.23", "1.23e-4" that can look like both numbers and strings, while discarding other crud.

## Related Issue
- Fixes #36037 

## How Has This Been Tested?
Local run of acceptance test and observe that the log message no longer happens.
```
make test-acceptance-webui BEHAT_FEATURE=tests/acceptance/features/webUISharingInternalGroups/shareWithGroups.feature:239
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

phil-davis committed:
Fix invalid foreach in SharingBlacklist.php

phil-davis commented on apps/files_sharing/lib/SharingBlacklist.php:
```diff
@@ -89,8 +93,7 @@ private function initCache() {
 	private function fetchBlacklistedReceiverGroupIds() {
-		$configuredBlacklist = $this->config->getAppValue('files_sharing', 'blacklisted_receiver_groups', '[]');
-		$decodedGroups = \json_decode($configuredBlacklist, true);
+		$decodedGroups = $this->getBlacklistedReceiverGroups();
```
I refactored here, because this code was duplicating what `getBlacklistedReceiverGroups` was already supposed to be doing.

jvillafanez replied:
I'd prefer to keep it like before. The reason is that this might mess up with the unittest code coverage.
The problem I see is that the public method `getBlacklistedReceiverGroups` will be called by a private method. This means that some code in the public method might not be tested while being marked as covered because of the private method's call.

phil-davis replied:
Or do I make a private `getBlacklistedReceiverGroupsXyz()` and call it from both `getBlacklistedReceiverGroups` and `fetchBlacklistedReceiverGroupIds` ?

jvillafanez replied:
Works for me. The original code was only one line, so the effort wasn't worthy.

jvillafanez commented on apps/files_sharing/lib/SharingBlacklist.php:
```diff
@@ -75,7 +75,11 @@ public function setBlacklistedReceiverGroups(array $ids) {
 	public function getBlacklistedReceiverGroups() {
-		return \json_decode($this->config->getAppValue('files_sharing', 'blacklisted_receiver_groups', '[]'), true);
+		$configuredBlacklist = $this->config->getAppValue('files_sharing', 'blacklisted_receiver_groups', '[]');
+		if ($configuredBlacklist === "") {
```
I'd prefer to fail the method, either returning false or throwing an exception, if the decoded json isn't an array. As things are with this code, you also need to check if the app value returns a string, a number or whatever random value gets there, not just an empty string.

The code assumed that the `setBlacklistedReceiverGroups` would be use, so it has guarantee that an array will be stored.

phil-davis replied:
Hmm - so somehow on my system I have ended up with the "empty list" being stored as the empty string, rather than the "empty JSON array".
I will see if I can make that happen for real using the webUI - if so, then that needs to be fixed anyway.

jvillafanez replied:
```
core/tests/acceptance/features/bootstrap/WebUIAdminSharingSettingsContext.php:
  353  		$this->featureContext = $environment->getContext('FeatureContext');
  354  		SetupHelper::runOcc(
  355: 			['config:app:set files_sharing blacklisted_receiver_groups --value=']
  356  		);
  357  	}
```
I guess that's the cause.

Taking into account that the admin can mess it up via occ, I think it's a good idea to harden the check.

PVince81 commented:
@phil-davis any update ?

phil-davis commented:
> @phil-davis any update ?

I thought it was going to be an easy/obvious fix - silly me. Now I need to spend some time "hardening" against this setting being potentially any invalid JSON or weird JSON.

phil-davis commented:
@jvillafanez @PVince81 I switched around the logic so that the code is in the private method `fetchBlacklistedReceiverGroupIds` rather than previously, where I had the private method calling the public method.

I added a more general check using `json_last_error` and report the problem as an error log. But I go on by setting the group ids empty.

Please review.

jvillafanez commented on apps/files_sharing/lib/SharingBlacklist.php:
```diff
@@ -88,10 +88,27 @@ private function initCache() {
 
-	private function fetchBlacklistedReceiverGroupIds() {
+	/**
+	 * @param bool $returnAnArrayWithGroupsAsValues if true then the group ids are the values in the array
```
Text needs to be fixed

jvillafanez commented on apps/files_sharing/lib/SharingBlacklist.php:
```diff
@@ -88,10 +88,27 @@ private function initCache() {
+		if (!\is_array($decodedGroups)) {
+			$decodedGroups = [];
+			$jsonError = \json_last_error();
+			\OC::$server->getLogger()->error('SharingBlacklist::getBlacklistedReceiverGroups - can\'t decode blacklisted_receiver_groups: ' . \print_r($configuredBlacklist, true) . ' - json error code: ' . $jsonError, ['app' => 'files_sharing']);
```
just ignore the error message. This needs to be injected in order to include unittests, which is too much work for little benefit. Adding a comment in the PHPDocs saying that an empty list will be returned if we can't parse the json is enough.
Note that there is no need for anyone to know that we're storing the information as a json string.

jvillafanez commented:
Note that this needs unittests to cover this new case.

phil-davis commented:
@jvillafanez or anybody - ready for review again.
I went crazy and covered whatever crappy things I thought someone might be able to set this to.

jvillafanez commented on apps/files_sharing/lib/SharingBlacklist.php:
```diff
@@ -88,12 +88,37 @@ private function initCache() {
+	 *                  then an empty list is returned.
+	 *                  Any parsed group ids that are not string, int or float are dropped from the list.
+	 */
+	private function fetchBlacklistedReceiverGroupIds($returnAnArrayWithGroupsAsValues = false) {
```
return the group set as originally planned, and use the `array_keys` function in the caller which needs to return the group set as list. This function shouldn't return different data

jvillafanez commented on apps/files_sharing/tests/SharingBlacklistTest.php:
```diff
@@ -63,8 +71,12 @@ public function testSetGetBlacklistedReceiverGroups($ids) {
+		if (\is_array($ids)) {
+			$this->sharingBlacklist->setBlacklistedReceiverGroups($ids);
+		} else {
+			$this->config->setAppValue('files_sharing', 'blacklisted_receiver_groups', $ids);
```
I think this is a different test. Although the call is mocked, I don't really like to have "direct" calls to objects other than the one being tested.
You can test the same with something like:
```
$this->config->method('getAppValue')->willReturn($ids);
$this->assertEquals($result, $this->sharingBlacklist->getBlacklistedReceiverGroups());
```
This seems a more standard approach.

phil-davis commented on apps/files_sharing/tests/SharingBlacklistTest.php:
```diff
@@ -40,7 +40,7 @@ public function setUp() {
 	}
 
-	public function setGetBlacklistedReceiverGroupsProvider() {
+	public function setBlacklistedReceiverGroupsProvider() {
```
This existing set of unit test cases is really about testing that `setBlacklistedReceiverGroups`  works.
"by accident" it uses `getBlacklistedReceiverGroups` to check the result.

jvillafanez replied:
Technically the test checks that you get what you set. It's intentional. I agree it isn't an unittest, but I don't think it's worthy to check that a oneline function call one specific method. I think the previous name make this more evident (or not)

phil-davis replied:
I put the original name back

phil-davis commented on apps/files_sharing/tests/SharingBlacklistTest.php:
```diff
@@ -67,6 +67,30 @@ public function testSetGetBlacklistedReceiverGroups($ids) {
 		$this->assertEquals($ids, $this->sharingBlacklist->getBlacklistedReceiverGroups());
 	}
 
+	public function getBlacklistedReceiverGroupsProvider() {
```
These test cases let us "manually" insert various strings as the blacklisted receiver groups config setting.
That happens in real life if the sysadmin "manually" sets the config entry with, for example, an `occ` command.

They check that for various "weird" strings of valid or invalid JSON the code does something reasonable.

jvillafanez replied:
If you think this is worthy to know, I think it's better to write a code comment. Personally, I'd rather not to do archeology to know why the test was created or any other important thing

phil-davis replied:
Description added to PHPdoc of method

phil-davis commented:
@jvillafanez I refactored stuff - please review again.

jvillafanez commented on apps/files_sharing/lib/SharingBlacklist.php:
```diff
@@ -88,12 +88,30 @@ private function initCache() {
+		// switch around so that the GroupIds are the keys of an array
+		$groupSet = [];
+		foreach ($validDecodedGroups as $group) {
 			$groupSet[$group] = true;
```
you can generate the group set directly during the first iteration of the groups.

phil-davis replied:
yes, good idea.

jvillafanez reviewed (commented):
one last thing and we're done

phil-davis commented:
@jvillafanez ready for review again

sharidas commented:
@phil-davis I am merging this change.

HanaGemela commented:
Cannot share to blacklisted group with group names like "0", "1", "1.23", "1.23e-4" on the desktop sync client 2.6.0 RC1, macOS 10.14.6 and testpilot 2.5.4
